src/viur/core/bones/phone.py

```diff
@@ -0,0 +1,119 @@
+
+
+class PhoneBone(StringBone):
+    """
+    The PhoneBone class is designed to store validated phone/fax numbers in configurable formats.
+    This class provides a number validation method, ensuring that the given phone/fax number conforms to the
+    required/configured format and structure.
+    """
+    type: str = "str.phone"
+    """
+    A string representing the type of the bone, in this case "str.phone".
+    """
+    def __init__(
+        self,
+        custom_regex: Optional[Pattern[str]] = None,
+        max_length: int = 15,
+        default_country_code: str = "+49",
+        apply_default_country_code: bool = False,
+        **kwargs: Any
+    ) -> None:
+        """
+        Initializes the PhoneBone with an optional custom regex for phone number validation, a default country code,
+        and a flag to apply the default country code if none is provided.
+
+        Args:
+            custom_regex (Optional[Pattern[str]]): An optional custom regex pattern for phone number validation.
+            max_length (int): The maximum length of the phone number.
+            default_country_code (str): The default country code to apply if none is provided.
+            apply_default_country_code (bool): Whether to apply the default country code if none is provided.
+            **kwargs (Any): Additional keyword arguments.
+        """
+        self.custom_regex: Pattern[str] = custom_regex or re.compile(
+            r'^(?:\+|00)?[1-9]\d{0,2}[-\s]?\d{1,4}[-\s]?\d{1,4}[-\s]?\d{1,4}$|'
+            r'^(0\d{1,4}[-\s]?\d{1,4}[-\s]?\d{1,4})$'
+        )
+        self.phone_regex: Pattern[str] = (
+            re.compile(self.custom_regex) if isinstance(self.custom_regex, str) else self.custom_regex
+        )
+        self.default_country_code: str = default_country_code
+        self.apply_default_country_code: bool = apply_default_country_code
+
+        super().__init__(max_length=max_length, **kwargs)
+
+    def isInvalid(self, value: str) -> Optional[str]:
+        """
+        Checks if the provided phone number is valid or not.
+
+        Args:
+            value (str): The phone number to be validated.
+
+        Returns:
+            Optional[str]: An error message if the phone number is invalid or None if it is valid.
+
+        The method checks if the provided phone number is valid according to the following criteria:
+        1. The phone number must not be empty.
+        2. The phone number must match the provided or default phone number format.
+        3. If the phone number has no country code and apply_default_country_code is True,
+           the default country code is applied.
+        """
+        if not value:
+            return "No value entered"
+
+        if not self.phone_regex.match(value):
+            return "Invalid phone number entered"
+
+        is_invalid: Optional[str] = super().isInvalid(value)
+
+        if is_invalid:
+            return is_invalid
+        return None
```

Comment:
You should add a super-call here to StringBones isValid-function. This allows further customization.

Reply:
i cant find a isValid function in StringBone neither in BaseBone, where is the definition of isValid ?
an ive added a super-call  to StringBone.isInvalid already

Reply:
I've meant `isInvalid` - sorry. One of ViUR's many ugly names...